src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3523,127 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+                if let Some(def) = name_binding.def() {
+                    if name == lookup_name && ns == namespace && filter_fn(def) {
+                        // create the path
+                        let ident = hir::Ident::from_name(name);
+                        let params = PathParameters::none();
+                        let segment = PathSegment {
+                            identifier: ident,
+                            parameters: params,
+                        };
+                        let span = name_binding.span.unwrap_or(syntax::codemap::DUMMY_SP);
+                        let mut segms = path_segments.clone();
+                        segms.push(segment);
+                        let segms = HirVec::from_vec(segms);
+                        let path = Path {
+                            span: span,
+                            global: true,
+                            segments: segms,
+                        };
+                        // push result
+                        // the entity is not accessible in the following cases:
+                        // 1. if it's defined in the same crate, it's always
+                        // accessible
+                        // 2. if it's defined in another crate, it's accessible
+                        // only if both the module is public and the entity is
+                        // declared as public
+                        lookup_results_everything.push(path.clone());
+                        if !in_module_is_extern ||
+                           name_binding.is_public() && in_module_is_accessible {
+                            lookup_results_accessible.push(path);
+                        }
+                    }
+                }
+
+                // collect submodules to explore
+                if let Some(module) = name_binding.module() {
+                    // form the path
+                    let path_segments = match module.parent_link {
+                        NoParentLink => path_segments.clone(),
+                        ModuleParentLink(_, name) => {
+                            let mut paths = path_segments.clone();
+                            let ident = hir::Ident::from_name(name);
+                            let params = PathParameters::none();
+                            let segm = PathSegment {
+                                identifier: ident,
+                                parameters: params,
+                            };
+                            paths.push(segm);
+                            paths
+                        }
+                        _ => unreachable!(),
+                    };
+
+                    // add the module to the lookup
+                    let is_public = in_module_is_accessible && name_binding.is_public();
+                    let is_extern = in_module_is_extern || module.is_extern_crate;
+                    worklist.push((module, path_segments, is_public, is_extern));
```

Comment:
We should only add to the worklist if `!in_module_is_extern || name_binding.is_public()`.